k2/csrc/ragged_ops_inl.h

```diff
@@ -71,23 +72,57 @@ void ApplyOpPerSublist(Ragged<T> &src, T default_value, Array1<T> *dst) {
     // the first time is to determine temporary device storage requirements
     K2_CUDA_SAFE_CALL(cub::DeviceSegmentedReduce::Reduce(
         nullptr, temp_storage_bytes, values_data, output_data, num_rows,
-        row_splits, row_splits + 1, op, default_value, c->GetCudaStream()));
+        row_splits, row_splits + 1, op, initial_value, c->GetCudaStream()));
     Array1<int8_t> d_temp_storage(c, temp_storage_bytes);
     K2_CUDA_SAFE_CALL(cub::DeviceSegmentedReduce::Reduce(
         d_temp_storage.Data(), temp_storage_bytes, values_data, output_data,
-        num_rows, row_splits, row_splits + 1, op, default_value,
+        num_rows, row_splits, row_splits + 1, op, initial_value,
         c->GetCudaStream()));
   }
 }
 
+template <typename T>
+Ragged<T> NormalizePerSublist(Ragged<T> &src) {
+  K2_STATIC_ASSERT(
+      (std::is_same<float, T>::value || std::is_same<double, T>::value));
+  T negative_infinity = -std::numeric_limits<T>::infinity();
+
+  ContextPtr &context = src.Context();
+  int32_t num_axes = src.NumAxes();
+  Array1<T> values(context, src.TotSize(num_axes - 2));
+  LogSumPerSublist(src, negative_infinity, &values);
+
+  const T *values_data = values.Data();
+  int32_t num_rows = values.Dim();
+  const int32_t *row_splits_data = src.RowSplits(num_axes - 1).Data();
+
+  Array1<T> ans_values(context, src.values.Dim());
+  Ragged<T> ans(src.shape, ans_values);
+
+  T *ans_data = ans.values.Data();
+  const T *src_data = src.values.Data();
+
+  K2_EVAL(
+      context, num_rows, lambda_do_normalization, (int32_t i)->void {
+        int32_t begin = row_splits_data[i];
+        int32_t end = row_splits_data[i + 1];
+        T normalizer = values_data[i];
+
+        // it also handles the case when the sublist is empty
+        for (int32_t k = begin; k != end; ++k)
```

Comment:
mm, @csukuangfj , I think @danpovey and I have said many times about this: avoid loop in kenel as best as you can...

Please use num_elements as the lambda index and get the index to `normalizer` with row_id.

Reply:
thanks, fixed.